tests/staggered_invert_test.cpp

```diff
@@ -1,692 +1,406 @@
+          printfQuda(" - level %d Chebyshev polynomial minumum %e\n", i + 1, mg_eig_amin[i]);
+          if (mg_eig_amax[i] <= 0)
+            printfQuda(" - level %d Chebyshev polynomial maximum will be computed\n", i + 1);
+          else
+            printfQuda(" - level %d Chebyshev polynomial maximum %e\n", i + 1, mg_eig_amax[i]);
+        }
+        printfQuda("\n");
+      }
+    }
+  }
 
-cpuColorSpinorField *in;
-cpuColorSpinorField *out;
-cpuColorSpinorField *ref;
-cpuColorSpinorField *tmp;
+  if (inv_deflate) {
+    printfQuda("\n   Eigensolver parameters\n");
+    printfQuda(" - solver mode %s\n", get_eig_type_str(eig_type));
+    printfQuda(" - spectrum requested %s\n", get_eig_spectrum_str(eig_spectrum));
+    printfQuda(" - number of eigenvectors requested %d\n", eig_nConv);
+    printfQuda(" - size of eigenvector search space %d\n", eig_nEv);
+    printfQuda(" - size of Krylov space %d\n", eig_nKr);
+    printfQuda(" - solver tolerance %e\n", eig_tol);
+    printfQuda(" - convergence required (%s)\n", eig_require_convergence ? "true" : "false");
+    if (eig_compute_svd) {
+      printfQuda(" - Operator: MdagM. Will compute SVD of M\n");
+      printfQuda(" - ***********************************************************\n");
+      printfQuda(" - **** Overriding any previous choices of operator type. ****\n");
+      printfQuda(" - ****    SVD demands normal operator, will use MdagM    ****\n");
+      printfQuda(" - ***********************************************************\n");
+    } else {
+      printfQuda(" - Operator: daggered (%s) , norm-op (%s)\n", eig_use_dagger ? "true" : "false",
+                 eig_use_normop ? "true" : "false");
+    }
+    if (eig_use_poly_acc) {
+      printfQuda(" - Chebyshev polynomial degree %d\n", eig_poly_deg);
+      printfQuda(" - Chebyshev polynomial minumum %e\n", eig_amin);
+      if (eig_amax <= 0)
+        printfQuda(" - Chebyshev polynomial maximum will be computed\n");
+      else
+        printfQuda(" - Chebyshev polynomial maximum %e\n\n", eig_amax);
+    }
+  }
 
-static void end();
+  printfQuda("Grid partition info:     X  Y  Z  T\n");
+  printfQuda("                         %d  %d  %d  %d\n", dimPartitioned(0), dimPartitioned(1), dimPartitioned(2),
+             dimPartitioned(3));
+}
 
-// Parameters defining the eigensolver
-void setEigParam(QudaEigParam &eig_param)
+int main(int argc, char **argv)
 {
-  eig_param.eig_type = eig_type;
-  eig_param.spectrum = eig_spectrum;
-  if ((eig_type == QUDA_EIG_TR_LANCZOS || eig_type == QUDA_EIG_IR_LANCZOS)
-      && !(eig_spectrum == QUDA_SPECTRUM_LR_EIG || eig_spectrum == QUDA_SPECTRUM_SR_EIG)) {
-    errorQuda("Only real spectrum type (LR or SR) can be passed to Lanczos type solver");
+  setQudaDefaultMgTestParams();
+  // Parse command line options
+  auto app = make_app();
+  add_eigen_option_group(app);
+  add_deflation_option_group(app);
+  add_multigrid_option_group(app);
```

Comment:
should this depend on whether QUDA is built with MG?

Reply:
Well, the `add_multigrid_option_group(app);` line will have to remain, albeit around preprocessor guards. I merged the `tests/staggered_invert_test.cpp` and `tests/staggered_multigrid_invert_test.cpp` and placed the relevant mg stuff around a run time evaluated boolean. The worst that can happen here is that a user invokes the help command, MG options are displayed, but MG is not built... which they will quickly discover if they run with `--inv-multigrid true`

Reply:
I like your idea of adding preprocessor guards @cpviolator , either here or perhaps better within the `add_multigrid_option_group` definition itself. It's a simple fix and could save the end user confusion.

Reply:
If we block out all the mg option group code from command_line_params.cpp, then I think it would cause even more confusion to the user.

If the user fails to build the multigrid code (which is defaulted not to build) and they try to pass **correct** mg options via the command line, then the user will be presented with an error message informing them that their mg options are unrecognised. This would be a huge red herring, given that every other unbuilt option in QUDA would still be recognised by the CLI parser.

On the other hand, as the code is presently, if the user passes valid mg options and mg is not built, they will be given sensible error messages at runtime.

Reply:
So is the solution adding a guard within `[staggered_]invert_test` instead? That's fine, I trust your judgement.

Reply:
Ok, I guess we can leave the options in. I think we already have a ton of options that are only caught at runtime later. @cpviolator can you file an issue so maybe we can look at that when someone has a good idea how to catch that when processing command line options...